Port the various FetchOptions to the new IPC serialization format

gavin-apple opened the pull request:
#### df928f2f3d5af2580922ca6b2584f0b611b85643
<pre>
Port the various FetchOptions to the new IPC serialization format
<a href="https://bugs.webkit.org/show_bug.cgi?id=249311">https://bugs.webkit.org/show_bug.cgi?id=249311</a>
rdar://103355724

Reviewed by Alex Christensen.

This change includes ports for:
    - FetchOptionsDestination
    - FetchOptionsMode
    - FetchOptionsCredentials
    - FetchOptionsCache
    - FetchOptionsRedirect
    - FetchOptions
    - FetchHeadersGuard

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/fetch/FetchHeaders.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/FetchOptions.h:
(WebCore::FetchOptions::encode const): Deleted.
(WebCore::FetchOptions::decode): Deleted.
* Source/WebCore/loader/FetchOptionsCache.h: Added.
* Source/WebCore/loader/FetchOptionsCredentials.h: Added.
* Source/WebCore/loader/FetchOptionsDestination.h: Added.
* Source/WebCore/loader/FetchOptionsMode.h: Added.
* Source/WebCore/loader/FetchOptionsRedirect.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.messages.in:

Canonical link: <a href="https://commits.webkit.org/257941@main">https://commits.webkit.org/257941@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/908e6be47c287bdcd77a5b2985915d5eb84dba71

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100403 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9565 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33467 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109719 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169965 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104396 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10476 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/131 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92806 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107597 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106182 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7930 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91198 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34581 "Passed tests") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22588 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3304 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24107 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3295 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9416 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43597 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5112 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2836 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/fetch/FetchHeaders.h:
```diff
@@ -38,20 +38,20 @@ namespace WebCore {
+
 class FetchHeaders : public RefCounted<FetchHeaders> {
 public:
-    enum class Guard {
```
Instead of removing it, we could do using Guard = FetchHeadersGuard.
That would remove the need of some of the changes

gavin-apple replied:
Thanks for the tip. PR is now **MUCH** cleaner.

gavin-apple committed:
Port the various FetchOptions to the new IPC serialization format
https://bugs.webkit.org/show_bug.cgi?id=249311
rdar://103355724

Reviewed by Alex Christensen.

This change includes ports for:
    - FetchOptionsDestination
    - FetchOptionsMode
    - FetchOptionsCredentials
    - FetchOptionsCache
    - FetchOptionsRedirect
    - FetchOptions
    - FetchHeadersGuard

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/fetch/FetchHeaders.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/FetchOptions.h:
(WebCore::FetchOptions::encode const): Deleted.
(WebCore::FetchOptions::decode): Deleted.
* Source/WebCore/loader/FetchOptionsCache.h: Added.
* Source/WebCore/loader/FetchOptionsCredentials.h: Added.
* Source/WebCore/loader/FetchOptionsDestination.h: Added.
* Source/WebCore/loader/FetchOptionsMode.h: Added.
* Source/WebCore/loader/FetchOptionsRedirect.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.messages.in:

Canonical link: https://commits.webkit.org/257941@main
